Improve ec2_ami tests

willthames opened the pull request:
##### SUMMARY

Ensure that ec2_ami_image_id fact gets set immediately after AMI creation so that they get torn down even if tests fail

Use YAML anchor to simplify AWS credential passing

Use aws_connection_info to reduce AWS credential boilerplate

Improve exception handling when updating image attributes

Error messages weren't correctly formatted to show image ids.


##### ISSUE TYPE
 - Bugfix Pull Request


##### COMPONENT NAME
ec2_ami

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.6.0 (devel e2aa1155ba) last updated 2018/04/19 10:02:37 (GMT +1000)
  config file = None
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible/lib/ansible
  executable location = /home/will/src/ansible/bin/ansible
  python version = 2.7.14 (default, Mar 14 2018, 13:36:31) [GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]
```


##### ADDITIONAL INFORMATION

maishsk commented:
@s-hertel I think there might be some confusion. 
I opened the original issue because there were some actions that were not working as in previous versions. 
Not all actions required a name parameter - I think that only actions actually created the image - need that parameter. 

In my use case - I was just modifying permissions on the image - not creating / deleting. 
In this use case - in 2.4 no `name` parameter was needed - but in 2.5 (after the move to boto3 - I think) it was mandatory.

I don't know if this makes things clearer - or more complicated - but I hope it provides context.

s-hertel commented:
Thanks, @maishsk. I'm not sure whether this pull request is fixing something. I believe my patch https://github.com/ansible/ansible/commit/e2aa1155ba5889296d2f018fdf8284c33f5f4918 has not impacted the use case of modifying an ami without a name and should keep the behavior you added in #38514 to fix the regression. If you're able to test that it would be very helpful.

willthames commented:
Thanks @s-hertel and @maishsk - maybe I completely misunderstood the original issue.

If #38514 still fixes the original issue, then I'll just update this PR to improve the integration tests (boilerplate and also the move of `set_fact` so that teardown works if tests fail)

Commit:
Improve ec2_ami tests

Ensure that ec2_ami_image_id fact gets set immediately after AMI
creation so that they get torn down even if tests fail

Use YAML anchor to simplify AWS credential passing

Use aws_connection_info to reduce AWS credential boilerplate

Improve exception handling when updating image attributes

Error messages weren't correctly formatted to show image ids.

willthames commented:
@s-hertel this is now just a tests improvement PR

s-hertel committed:
provide name for image creation
